src/charm.py

```diff
@@ -43,6 +45,71 @@ def _aggregate_alerts(rules: Dict, rule_path_map: RulesMapping, forward_alert_ru
         logger.debug(f"updated alert rules file {rule_file.as_posix()}")
 
 
+def dashboards(relations: List[Relation]) -> list:
+    """Returns an aggregate of all dashboards received by this otelcol."""
+    aggregate = {}
+    for rel in relations:
+        dashboards = json.loads(rel.data[rel.app].get("dashboards", "{}"))  # type: ignore
+        if "templates" not in dashboards:
+            continue
+        for template in dashboards["templates"]:
+            content = json.loads(
+                LZMABase64.decompress(dashboards["templates"][template].get("content"))
+            )
+            entry = {
+                "charm": dashboards["templates"][template].get("charm", "charm_name"),
+                "relation_id": rel.id,
+                "title": template,
+                "content": content,
+            }
+            aggregate[template] = entry
+
+    return list(aggregate.values())
+
+
+def forward_dashboards(charm: CharmBase):
+    """Instantiate the GrafanaDashboardProvider and update the dashboards in the relation databag.
+
+    First, dashboards from relations (including those bundled with Otelcol) and save them to disk.
+    Then, update the relation databag with these dashboards for Grafana.
+    """
+    charm_root = charm.charm_dir.absolute()
+    dashboard_paths = PathMapping(
+        src=charm_root.joinpath(*"src/grafana_dashboards".split("/")),
+        dest=charm_root.joinpath(*"grafana_dashboards".split("/")),
+    )
+    if not os.path.isdir(dashboard_paths.dest):
+        shutil.copytree(dashboard_paths.src, dashboard_paths.dest, dirs_exist_ok=True)
+    grafana_dashboards_provider = GrafanaDashboardProvider(
+        charm,
+        relation_name="grafana-dashboards-provider",
+        dashboards_path=dashboard_paths.dest,
+    )
+
+    # Copy dashboards from relations and save them to disk."""
+    if not charm.unit.is_leader():
+        return
+    shutil.rmtree(dashboard_paths.dest)
+    shutil.copytree(dashboard_paths.src, dashboard_paths.dest)
+    for dash in dashboards(charm.model.relations["grafana-dashboards-consumer"]):
+        # Build dashboard custom filename
+        charm = dash.get("charm", "charm-name")
+        rel_id = dash.get("relation_id", "rel_id")
+        title = dash.get("title").replace(" ", "_").replace("/", "_").lower()
+        filename = f"juju_{title}-{charm}-{rel_id}.json"
+        with open(Path(dashboard_paths.dest, filename), mode="w", encoding="utf-8") as f:
+            f.write(json.dumps(dash["content"]))
+            logger.debug("updated dashboard file %s", f.name)
+
+    # Scan the built-in dashboards and update relations with changes
+    # TODO Make this a public method like reload_alerts for logs and metrics
+    grafana_dashboards_provider._update_all_dashboards_from_dir()
```

Comment:
Can we create the `GrafanaDashboardProvider` instance immediately before this line?